inngest-core/src/main/kotlin/io/inngest/Function.kt

```diff
@@ -33,15 +33,31 @@ enum class ResultStatusCode(val code: Int, val message: String) {
     Error(500, "Function Error"),
 }
 
-data class StepResult(
-    val data: Any? = null,
+
+abstract class StepOp(
     // The hashed ID of a step
-    val id: String = "",
-    val name: String = "",
-    val op: OpCode,
-    val statusCode: ResultStatusCode,
+    open val id: String = "",
+    open val name: String = "",
+    open val op: OpCode,
+    open val statusCode: ResultStatusCode,
 )
 
+data class StepResult(
+    override val id: String,
+    override val name: String,
+    override val op: OpCode,
+    override val statusCode: ResultStatusCode,
+    val data: Any? = null,
+) : StepOp(id, name, op, statusCode)
+
+data class StepOptions(
+    override val id: String,
+    override val name: String,
+    override val op: OpCode,
+    override val statusCode: ResultStatusCode,
+    val opts: HashMap<String, String>? = null,
+) : StepOp(id, name, op, statusCode)
```

Comment:
Is the difference here that we're _reporting_ a step instead of telling the Inngest Server that we've run a step? The spec could be clearer on that terminology.

If so, it'd be nice to call this `StepPlan` or `PlannedStep`.

Reply:
🤔 Not quite sure about that. I quickly decided to call it `StepOp` to say that a step operation completed. Some contain `data` (`StepResult`) and some contain `opts` (`StepOptions`) - Maybe they're bad names, but I wanted to use an abstract class and inherit from it so there isn't one large data class with a bunch of optional parameters. This felt more type safe. I may be wrong!

Reply:
@djfarrelly My comment might have targetted the wrong lines. 😆 I mean `StepOptions -> StepPlan`. Regardless, you bring up an interesting point. We've always returned the entire step, e.g.:

```ts
Array<{
	id: string;
	op: string;
	displayName?: string;
	opts?: { [key: string]: any };
}>
```

Then we **also** include `name`, `data`, and `error` as they're required. You're saying that when returning the result of a step, you drop the fields you don't need (because they've already been reported), like `opts`.

I'd err on the side of caution and still provide the entire step. For example, when returning an `error`, `opts` will likely later return a retry count that the Inngest Server will use.

I don't think there are any ill effects of this currently. I'll have to think on it more and we can optimize the size of the data going back and forth. It's hard to say that we'll never need it, though.